File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.py
import sys
from collections import deque


def bfs(start):
    dist = [0] * (n + 1)

    q = deque()
    q.append(start)

    while q:
        x = q.popleft()
        visited[x] = True

        for i in graph[x]:
            if not visited[i]:
                visited[i] = True
                dist[i] = dist[x] + 1
                q.append(i)
    return sum(dist)


n, m = map(int, sys.stdin.readline().split())

result = 0

graph = [[] for _ in range(n + 1)]

for _ in range(m):
    a, b = map(int, sys.stdin.readline().split())
    graph[a].append(b)
    graph[b].append(a)

min_dist = int(1e9)

for i in range(1, n + 1):
    visited = [False] * (n + 1)
    dist = bfs(i)

    if min_dist > dist:
        min_dist = dist
        result = i
print(result)

File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[][] graph = new int[n + 1][n + 1];
        for (int i = 1; i <= n; i++) {
            Arrays.fill(graph[i], (int) 1e9);
            graph[i][i] = 0;
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            graph[a][b] = 1;
            graph[b][a] = 1;
        }

        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]);
                }
            }
        }

        int result = 101;
        int dist = (int) 1e9;

        for (int i = 1; i <= n; i++) {
            int sum = 0;
            for (int j = 1; j <= n; j++) {
                sum += graph[i][j];
            }
            if (sum < dist) {
                result = i;
                dist = sum;
            }
        }

        System.out.println(result);
    }
}
